File: docs-site/generators/homepage.js
const fs = require("fs");
const path = require("path");
const yaml = require("js-yaml");

class HomepageGenerator {
  constructor(config) {
    this.config = config;
  }

  generate() {
    console.log("Generating homepage from README...");

    try {
      const readmePath = path.join("..", "README.md");
      const readmeContent = fs.readFileSync(readmePath, "utf8");

      // Extract content after the first heading
      const lines = readmeContent.split("\n");
      let contentStart = 0;
      for (let i = 0; i < lines.length; i++) {
        if (lines[i].startsWith("# ")) {
          contentStart = i;
          break;
        }
      }

      let content = lines.slice(contentStart).join("\n");

      // Fix links for Hugo with baseURL subdirectory
      // Convert docs-site/content/file.md -> file/ (relative format)
      content = content.replace(/docs-site\/content\/([^)]+)\.md/g, "$1/");

      // Convert any remaining .md links to Hugo format (relative)
      content = content.replace(/\]\(([^)]+)\.md\)/g, "]($1/)");

      // Fix docs-site root links
      content = content.replace(
        /\[([^\]]+)\]\(docs-site\/\)/g,
        '[$1]({{< ref "/" >}})',
      );

      // Fix LICENSE link
      content = content.replace(
        /\[([^\]]+)\]\(LICENSE\)/g,
        "[$1](https://github.com/otto-nation/otto-stack/blob/main/LICENSE)",
      );

      const today = new Date().toISOString().split("T")[0];
      const frontmatter = {
        title: "otto-stack",
        description:
          "A powerful development stack management tool built in Go for streamlined local development automation",
        lead: "Streamline your local development with powerful CLI tools and automated service management",
        date: "2025-10-01",
        lastmod: today,
        draft: false,
        weight: 50,
        toc: true,
      };

      const frontmatterYaml = yaml.dump(frontmatter);
      return `---\n${frontmatterYaml}---\n\n${content}`;
    } catch (error) {
      console.warn("Could not generate homepage:", error.message);
      return "";
    }
  }
}

module.exports = HomepageGenerator;

File: cmd/generate-docs/homepage.go
package main

import (
	"fmt"
	"os"
	"regexp"
	"strings"
)

func generateHomepage() error {
	data, err := os.ReadFile(readmePath)
	if err != nil {
		return fmt.Errorf("read README.md: %w", err)
	}

	content := extractFromFirstHeading(string(data))
	content = fixHomepageLinks(content)

	out, err := formatDocument(pageFM("homepage"), content)
	if err != nil {
		return err
	}
	return writeOutput(pageOutput("homepage"), out)
}

// extractFromFirstHeading trims content before the first top-level heading.
func extractFromFirstHeading(content string) string {
	lines := strings.Split(content, "\n")
	for i, line := range lines {
		if strings.HasPrefix(line, "# ") {
			return strings.Join(lines[i:], "\n")
		}
	}
	return content
}

// fixHomepageLinks rewrites README links to work under the Hugo site's baseURL.
func fixHomepageLinks(content string) string {
	// Convert docs-site/content/file.md → file/
	content = regexp.MustCompile(`docs-site/content/([^)]+)\.md`).
		ReplaceAllString(content, "$1/")

	// Convert remaining .md links to Hugo directory format
	content = regexp.MustCompile(`\]\(([^)]+)\.md\)`).
		ReplaceAllString(content, "]($1/)")

	// Convert docs-site root links to Hugo ref shortcode
	reSiteRoot := regexp.MustCompile(`\[([^\]]+)\]\(docs-site/\)`)
	content = reSiteRoot.ReplaceAllStringFunc(content, func(match string) string {
		inner := reSiteRoot.FindStringSubmatch(match)
		if len(inner) > 1 {
			return fmt.Sprintf(`[%s]({{< ref "/" >}})`, inner[1])
		}
		return match
	})

	// Expand bare LICENSE link to full GitHub URL
	content = regexp.MustCompile(`\[([^\]]+)\]\(LICENSE\)`).
		ReplaceAllString(content, "[$1](https://github.com/otto-nation/otto-stack/blob/main/LICENSE)")

	return content
}
